exec/java-exec/src/main/java/org/apache/drill/exec/rpc/security/plain/PlainFactory.java

```diff
@@ -0,0 +1,166 @@
+      if (cause instanceof LoginException) {
+        throw new SaslException("Failed to login.", cause);
+      }
+      throw new SaslException("Unexpected failure trying to login. ", cause);
+    }
+  }
+
+  @Override
+  public SaslServer createSaslServer(final UserGroupInformation ugi, final Map<String, ?> properties)
+      throws SaslException {
+    return FastSaslServerFactory.getInstance().createSaslServer(SIMPLE_NAME, null /** protocol */,
+        null /** serverName */, properties, new PlainServerCallbackHandler());
+  }
+
+  @Override
+  public SaslClient createSaslClient(final UserGroupInformation ugi, final Map<String, ?> properties)
+      throws SaslException {
+    final String userName = (String) properties.get(DrillProperties.USER);
+    final String password = (String) properties.get(DrillProperties.PASSWORD);
+
+    return FastSaslClientFactory.getInstance().createSaslClient(new String[]{SIMPLE_NAME},
+        null /** authorization ID */, null, null, properties, new CallbackHandler() {
+          @Override
+          public void handle(final Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+            for (final Callback callback : callbacks) {
+              if (callback instanceof NameCallback) {
+                NameCallback.class.cast(callback).setName(userName);
+                continue;
+              }
+              if (callback instanceof PasswordCallback) {
+                PasswordCallback.class.cast(callback).setPassword(password.toCharArray());
+                continue;
+              }
+              throw new UnsupportedCallbackException(callback);
+            }
+          }
+        });
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (authenticator != null) {
+      authenticator.close();
+    }
+  }
+
+  private class PlainServerCallbackHandler implements CallbackHandler {
+
+    @Override
+    public void handle(Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+      NameCallback nameCallback = null;
+      PasswordCallback passwordCallback = null;
+      AuthorizeCallback authorizeCallback = null;
+      for (final Callback callback : callbacks) {
+        if (callback instanceof NameCallback) {
+          nameCallback = NameCallback.class.cast(callback);
+        } else if (callback instanceof PasswordCallback) {
+          passwordCallback = PasswordCallback.class.cast(callback);
+        } else if (callback instanceof AuthorizeCallback) {
+          authorizeCallback = AuthorizeCallback.class.cast(callback);
+        } else {
+          throw new UnsupportedCallbackException(callback);
+        }
+      }
+
+      if (nameCallback == null || passwordCallback == null || authorizeCallback == null) {
+        throw new SaslException("Insufficient credentials.");
+      }
+
+      try {
+        assert authenticator != null;
+        authenticator.authenticate(nameCallback.getName(), new String(passwordCallback.getPassword()));
```

Comment:
I believe this is because `PlainFactory` relies on `FastSasl*Factory` which in turn uses `javax.security.sasl.Sasl` java class, which instantiates factories using the default constructor.

but you could directly instantiate `PlainServer` from `PlainFactory` without using reflection too (if someone where to use `Sasl#createSaslServer` directly, it would not work either because you also need to use the right callback class...)

I don't think the `Closable` is an issue: objects returned by PlainFactory should be considered invalid if the factory is closed (it's already the case).